### PR TITLE
Allocation engine

### PR DESCRIPTION
**Pull request by douglasrlee**
## Summary

Phase 3a of the Simple-style Expenses feature. As each scheduled paycheck date arrives, the engine fills expense buckets so that by the due date the bucket equals the expense amount — Simple's look-ahead split (remaining ÷ paychecks-remaining-until-due).

**Critical constraint — virtual money must be backed by real money.** A funding event *proposes* allocations on its scheduled date, but each allocation only "funds" (sets `funded_at`) when the user's available bank balance can back the running total. Funding happens in **due-date order** — most urgent expenses are funded first. Anything left pending stays pending and retries hourly until the balance catches up.

## What's in this PR

- **New models**: `FundingEvent` (per schedule per occurrence) and `Allocation` (per funding event per expense, with `funded_at` lifecycle).
- **`AllocationEngine` service**:
  - `propose_for(event)` creates pending `Allocation` rows (one per expense on the schedule).
  - `fund_pending_for(user)` flips `funded_at` on pending allocations in due-date order, walking the user's bank `available_balance`.
- **`ProcessFundingEventsJob`** (hourly via `config/recurring.yml`): materialize new events → propose allocations → fund pending allocations across all users.
- **`Expense#bucket_balance`** counts only funded allocations. **`Expense#off_track?`** is true when any allocation is still pending.
- **Expense row UI** gains a daisyUI progress bar (funded / target) plus an "Off track" badge when pending allocations exist.

## What's out of scope (later PRs)

- Phase 3b — transaction-to-expense linking (bucket reset on transaction match + due-date advance).
- Phase 3c — safe-to-spend on `/transactions` (`available_balance − Σ(bucket_balance)`).
- Manual "process now" button + Today view of in-flight + off-track allocations.

## Test plan

- [ ] Deploy. `heroku run rails db:migrate` succeeds.
- [ ] In the Mission Control UI at `/jobs`, confirm the new `process_funding_events` recurring entry shows up.
- [ ] Create a biweekly funding schedule starting today and two monthly expenses ($15.99 and $100, both tied to it).
- [ ] Trigger `ProcessFundingEventsJob.perform_now`. `/expenses` shows progress bars for each expense at the correct % and no "Off track" badges.
- [ ] Drop `available_balance` on the linked bank so only the more-urgent expense's allocation fits. Move time forward to the next event, re-run the job. The urgent expense funds; the other shows the warning-tinted bar + "Off track" badge.
- [ ] Bring `available_balance` back up. Next hourly run flips the pending allocation to funded; badge disappears.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/models/funding_schedule.rb [hunk 1]
```diff
@@ -17,6 +18,19 @@ class FundingSchedule < ApplicationRecord
...
+  def materialize_events_through(end_date:)
+    last = funding_events.maximum(:occurs_on)
+    cursor = last ? next_occurrences(count: 1, after: last + 1).first : start_date
+    events = []
+    while cursor && cursor <= end_date
+      events << funding_events.create_or_find_by!(occurs_on: cursor)
+      cursor = next_occurrences(count: 1, after: cursor + 1).first
+    end
+    events
+  end
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/models/funding_schedule.rb:33]: Good catch — switched to walking via the existing `advance()` helper so each step is O(1). Now O(n) overall.

File: app/services/allocation_engine.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                        .where(expenses: { user_id: user.id }, funded_at: nil)
+                        .order('expenses.due_on ASC, expenses.created_at ASC')
+
+    pending.find_each do |allocation|
+      next unless available >= already_funded + allocation.amount
+
+      allocation.update!(funded_at: Time.current)
+      already_funded += allocation.amount
+    end
+  end
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/services/allocation_engine.rb:49]: Real bug — switched to `.each` so the `ORDER BY expenses.due_on` is honored (pending allocations per user are bounded, so loading them all is fine), and wrapped the whole pass in `user.with_lock` so concurrent runs can't both read the same `already_funded` baseline and over-fund. Comment in the source notes why `find_each` was wrong here.

File: app/views/expenses/_expense.html.erb [hunk 3]
```diff
@@ -15,6 +15,17 @@
...
+      <div class="flex items-center gap-2 mt-1">
+        <progress class="progress <%= expense.off_track? ? 'progress-warning' : 'progress-primary' %> h-1.5 flex-1"
+                  value="<%= percent %>" max="100"></progress>
+        <span class="text-xs text-base-content/60 tabular-nums shrink-0">
+          <%= number_to_currency(expense.bucket_balance) %> / <%= number_to_currency(expense.amount) %>
+        </span>
+        <% if expense.off_track? %>
+          <span class="badge badge-warning badge-xs shrink-0">Off track</span>
+        <% end %>
+      </div>
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/views/expenses/_expense.html.erb]: Hoisted both into locals at the top of the row block. Sets us up for an eager-load follow-up (sum-of-allocations + has-pending preload on Expense) once we see real N+1 cost.

File: app/services/allocation_engine.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    funding_event.update!(processed_at: Time.current)
+  end
+
+  # Walk this user's pending allocations in due-date order and flip
+  # funded_at when current available_balance can back them. Serialized
+  # per-user with a row lock so concurrent runs can't double-fund off
+  # the same already_funded baseline.
+  def self.fund_pending_for(user)
+    user.with_lock do
+      available = user.bank_accounts.sum(:available_balance)
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/services/allocation_engine.rb]: Good catch — coalesced `user.bank_accounts.sum(:available_balance) || 0` with an inline comment noting that PG returns NULL when every input row is NULL (vs AR returning 0 for no rows).

File: app/jobs/process_funding_events_job.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    FundingEvent.where(processed_at: nil)
+                .where(occurs_on: ..Date.current)
+                .find_each do |event|
+      AllocationEngine.propose_for(event)
+    rescue StandardError => error
+      Rails.logger.error("AllocationEngine.propose_for failed for funding_event=#{event.id}: #{error.message}")
+      Appsignal.send_error(error)
+    end
+
+    User.joins(:expenses).distinct.find_each do |user|
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/jobs/process_funding_events_job.rb]: Narrowed the scan: `User.joins(expenses: :allocations).where(allocations: { funded_at: nil }).distinct`. Only users with at least one pending allocation get walked + locked. The propose phase above this still runs first, so allocations proposed earlier in the same job run are included.

File: app/services/allocation_engine.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  # Look-ahead split: remaining-amount / paychecks-remaining-until-due.
+  # Past-due falls to "this paycheck covers the rest" (clamped to 1).
+  private_class_method def self.compute_proposed_amount(expense, schedule, as_of:)
+    remaining = expense.amount - expense.allocations.sum(:amount)
+    return 0 if remaining <= 0
+
+    next_due = expense.next_due_on(after: as_of)
+    paychecks_remaining = schedule.next_occurrences(count: 100, after: as_of)
+                                  .count { |d| d <= next_due }
+    paychecks_remaining = 1 if paychecks_remaining.zero?
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/services/allocation_engine.rb:63]: Real bug — fixed. Added `FundingSchedule#occurrence_count_between(after:, through:)` which walks via the existing `advance()` helper for an accurate unbounded count, and swapped the call site to use it. Added a test that asserts 105 paychecks across a 2-year window on a weekly schedule (the old `count: 100` cap would have undercounted).

File: app/models/funding_schedule.rb [hunk 7]
```diff
@@ -17,6 +18,21 @@ class FundingSchedule < ApplicationRecord
 
   def semimonthly? = cadence == 'semimonthly'
 
+  # Find-or-creates a FundingEvent for every occurrence between the last
+  # materialized event (or start_date) through end_date. Idempotent.
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/models/funding_schedule.rb]: Fixed: "Finds or creates".

File: app/jobs/process_funding_events_job.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+  def perform
+    FundingSchedule.find_each do |schedule|
+      schedule.materialize_events_through(end_date: Date.current)
+    end
+
+    FundingEvent.where(processed_at: nil)
+                .where(occurs_on: ..Date.current)
+                .find_each do |event|
+      AllocationEngine.propose_for(event)
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/jobs/process_funding_events_job.rb:17]: Same root cause as the earlier `fund_pending_for` fix — `find_each` ignores ORDER BY. Switched to `.each` with explicit `.order(:occurs_on, :created_at)` so events get proposed oldest-first. Comment in the source notes why find_each was wrong here.

File: app/services/allocation_engine.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        next unless available >= already_funded + allocation.amount
+
+        allocation.update!(funded_at: Time.current)
+        already_funded += allocation.amount
+      end
+    end
+  end
+
+  # Look-ahead split: remaining-amount / paychecks-remaining-until-due.
+  # Past-due falls to "this paycheck covers the rest" (clamped to 1).
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/services/allocation_engine.rb]: Comment was misleading — when the expense is past-due, `next_due_on` actually rolls forward past `as_of`, so the normal split applies. The clamp-to-1 only fires when the schedule has no occurrences in `[as_of, next_due]` at all (e.g., schedule hasn't started yet). Updated the doc to describe what the code actually does.

File: app/models/funding_schedule.rb [hunk 10]
```diff
@@ -17,6 +18,36 @@ class FundingSchedule < ApplicationRecord
...
+  def materialize_events_through(end_date:)
+    return [] if start_date.blank? || cadence.blank?
+
+    last = funding_events.maximum(:occurs_on)
+    cursor = last ? advance(last) : start_date
+    events = []
+    while cursor && cursor <= end_date
+      events << funding_events.create_or_find_by!(occurs_on: cursor)
+      cursor = advance(cursor)
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/models/funding_schedule.rb]: Real edge case. Extracted `first_unmaterialized_occurrence` to clamp to `start_date` so we never backfill into the user's revised window. Test in `funding_schedule_test.rb` covers: create older events, move start_date forward, materialize — only the new-window event is created, no backfill into the skipped range.

File: app/views/expenses/_expense.html.erb [hunk 11]
```diff
@@ -15,6 +15,21 @@
...
         <% end %>
       </p>
+      <%
+        bucket = expense.bucket_balance
+        off_track = expense.off_track?
+        percent = expense.amount.positive? ? [ (bucket / expense.amount * 100).to_i, 100 ].min : 0
+      %>
+      <div class="flex items-center gap-2 mt-1">
+        <progress class="progress <%= off_track ? 'progress-warning' : 'progress-primary' %> h-1.5 flex-1"
+                  value="<%= percent %>" max="100"></progress>
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/views/expenses/_expense.html.erb]: Added `aria-label="${expense.name} funded ${percent}%"` so the progress is announced with context.
